src/viewer.js

```diff
@@ -2541,11 +2542,31 @@ function onCanvasDrag( event ) {
         if( !this.panVertical ){
             event.delta.y = 0;
         }
-        this.viewport.panBy( this.viewport.deltaPointsFromPixels( event.delta.negate() ), gestureSettings.flickEnabled );
+
         if( this.constrainDuringPan ){
-            this.viewport.applyConstraints();
+            var delta = this.viewport.deltaPointsFromPixels( event.delta.negate() );
+
+            this["viewport"].centerSpringX.target.value += delta.x;
+            this["viewport"].centerSpringY.target.value += delta.y;
+
+            var bounds = this.viewport.getBounds();
+            var constrainedBounds = this.viewport.getConstrainedBounds();
+
+            this["viewport"].centerSpringX.target.value -= delta.x;
+            this["viewport"].centerSpringY.target.value -= delta.y;
```

Comment:
Instead of changing and reverting the spring, could `getConstrainedBounds` take the delta as parameter?
Or maybe we could have a `viewport.adjustDeltaForConstraints(delta)` method which will do all of that?